study.py

```diff
@@ -5,31 +5,34 @@
 
 
 def study():
-    sim = Simulator(0)
+    config = Configuration(0, 64, 256, 0, 100, 20, 8, 16, 4, 0, 10, 1)
+    sim = Simulator(config)
     sim.initLogger()
-    maxTries = 10
-    step = 20
     frRange = []
+    results = []
     resultRange = []
     simCnt = 0
     sim.logger.info("Starting simulations:", extra=sim.format)
     start = time.time()
-    for fr in range(0, 100, step):
-        if fr % 10 == 0:
-            sim.logger.info("Failure rate %d %% ..." % fr, extra=sim.format)
+    for fr in range(config.failureRateStart, config.failureRateStop+1, config.failureRateStep):
         sim.resetFailureRate(fr)
-        result = 0
-        for i in range(maxTries):
-            sim.initValidators()
-            sim.initNetwork()
-            result += sim.run()
-            simCnt += 1
+        for chi in range(config.chiStart, config.chiStop+1, config.chiStep):
+            sim.resetChi(chi)
+            blockAvailable = 0
+            for run in range(config.runStart, config.runStop, config.runStep):
+                sim.logger.info("FR: %d %%, Chi: %d %%, Run: %d ..." % (fr, chi, run), extra=sim.format)
+                sim.initValidators()
+                sim.initNetwork()
+                result = sim.run()
+                blockAvailable += result.blockAvailable
+                results.append(result)
+                simCnt += 1
```

Comment:
I think I would prefer calling a Configuration what is processed by a single run, and use a list of configurations or a generator function to drive the study. That means chi and run would be part of the configuration, while chiStart/chiStop/chiStep would be one possible implementation of parameter space exploration. If you think it sounds good we can change it like that, but it is also OK like it is now, just a bit less generic.

Reply:
I ended up creating a new class called Shape that stores the "shape" of the current simulation, in other words, the exact parameters used in the current simulation. For runs with multiple simulations, I added a function to reset the shape with the new parameters. The configuration is now obtained from a configuration file that is passed by a parameter.

Reply:
Thanks, sounds great!